File: src/06.hashMap/19.350.go
package hashmap

import (
	"sort"
)

func Intersect(nums1 []int, nums2 []int) []int {
	// m, ans := make(map[int]int), []int{}

	// for _, v := range nums1 {
	// 	m[v] += 1
	// }

	// for _, v := range nums2 {
	// 	if n, ok := m[v]; ok && n > 0 {
	// 		m[v] = n - 1
	// 		ans = append(ans, v)
	// 	}
	// }

	// return ans

	// 如果给定的数组已经排好序呢？你将如何优化你的算法？
	ans, l, r := []int{}, 0, 0

	sort.Ints(nums1)
	sort.Ints(nums2)

	for l < len(nums1) && r < len(nums2) {
		if nums1[l] < nums2[r] {
			l++
		} else if nums1[l] > nums2[r] {
			r++
		} else {
			ans = append(ans, nums1[l])
			l++
			r++
		}
	}

	return ans
}

File: src/06.hashMap/19.350.ts
export function intersect(nums1: number[], nums2: number[]): number[] {
    // 正常解法
    // const map = new Map<number, number>();
    // const res: number[] = [];
    // for (let i = 0; i < nums1.length; i++) {
    //     map[nums1[i]] = (map[nums1[i]] || 0) + 1;
    // }
    // for (let i = 0; i < nums2.length; i++) {
    //     if (map[nums2[i]]) {
    //         res.push(nums2[i]);
    //         map[nums2[i]] = map[nums2[i]] - 1;
    //     }
    // }
    // return res;

    // 如果给定的数组已经排好序呢？你将如何优化你的算法？
    nums1.sort((a, b) => a - b);
    nums2.sort((a, b) => a - b);

    // 双指针
    let l = 0,
        r = 0;
    const ans: number[] = [];
    while (l < nums1.length && r < nums2.length) {
        if (nums1[l] < nums2[r]) {
            l++;
        } else if (nums1[l] === nums2[r]) {
            ans.push(nums1[l]);
            l++;
            r++;
        } else {
            r++;
        }
    }

    return ans;
}
